### PR TITLE
(un)linkapps: modernize, prune: remove broken app symlinks

### PR DESCRIPTION
**Pull request by UniqMartin**
The first two commits are all about using `Pathname` as much as possible (bringing the code up to current Homebrew standards), reducing duplication, and avoiding external commands in favor of native methods.

The third (and last) commit implements `brew unlinkapps --prune` and allows to remove broken symlinks from `/Applications` and `~/Applications` if they were previously created with `brew linkapps`. This part also provides `Homebrew.unlinkapps_prune` that could be used from `cmd/uninstall.rb` et al. to implement automatic pruning of broken symlinks as suggested in #46542 (will be addressed in a separate PR).

### CODE REVIEW

File: Library/Homebrew/cmd/linkapps.rb [hunk 1]
```diff
@@ -1,22 +1,21 @@
...
 
     if ARGV.named.empty?
-      kegs = Formula.racks.map do |rack|
-        keg = rack.subdirs.map { |d| Keg.new(d) }
-        next if keg.empty?
-        keg.detect(&:linked?) || keg.max { |a, b| a.version <=> b.version }
+      kegs = Formula.installed.map do |formula|
+        formula_kegs = formula.installed_kegs
+        next if formula_kegs.empty?
+        formula_kegs.detect(&:linked?) || formula_kegs.max_by(&:version)
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/linkapps.rb]: We shouldn't use `Formula.installed` here. The thing is commands like `link/unlink/switch/uninstall/linkapp` should work fine with DIY installation, i.e. no formula required. So let's use rack directly instead of trying to load formula.

Comment (UniqMartin) [inline Library/Homebrew/cmd/linkapps.rb]: Thanks for pointing this out! I'll fix it shortly. DIY seems like such an esoteric feature that I tend to always forget about it (and I never used it).

Comment (xu-cheng) [inline Library/Homebrew/cmd/linkapps.rb]: Beside the complete DIY installation(i.e. manually compiling while set prefix to cellar), another common situation is `brew install <URL/path to formula>` following with cache purge. Therefore, formula file will also be gone.

**Comment by mikemcquaid** [comment]
Looks good to me :+1: 

> This part also provides Homebrew.unlinkapps_prune that could be used from cmd/uninstall.rb et al. to implement automatic pruning of broken symlinks as suggested in #46542 (will be addressed in a separate PR).

That would be great. I was going to take a look but you're already closer to it :wink:

**Comment by UniqMartin** [comment]
I addressed the problem raised by @xu-cheng (thanks!) with the nice side-effect of being much faster now. Nothing else changed in what I just pushed.

> That would be great. I was going to take a look but you're already closer to it :wink:

I intended to do this anyway and as this issue of stale symlinks has resurfaced again in #46542, I decided to address it sooner than later. :smile_cat:

**Comment by UniqMartin** [comment]
Updated to address a merge conflict. It's :white_check_mark: again.

File: Library/Homebrew/manpages/brew.1.md [hunk 2]
```diff
@@ -451,6 +451,10 @@ With `--verbose` or `-v`, many commands print extra debugging information. Note
 
     If no <formulae> are provided, all linked app will be removed.
 
+  * `unlinkapps --prune`:
+    Removes links created by `brew linkapps` that are no longer valid, e.g.,
+    because a formula was removed or apps provided by a formula have changed.
```
Comment (mikemcquaid) [inline Library/Homebrew/manpages/brew.1.md]: Wondering if this behaviour should just be done by `brew prune`?

Comment (UniqMartin) [inline Library/Homebrew/manpages/brew.1.md]: This sounds like a better place for this functionality; thanks for pointing it out! Do you mean as default behavior for `brew prune` or as a separate mode of operation, e.g. `brew prune --apps`?

Comment (mikemcquaid) [inline Library/Homebrew/manpages/brew.1.md]: Default behaviour; I don't see a reason why you'd _not_ want to do this.

Comment (UniqMartin) [inline Library/Homebrew/manpages/brew.1.md]: Agreed; I'll update the code to move the new functionality into `brew prune`.

**Comment by UniqMartin** [comment]
Relocated pruning of broken app symlinks into `brew prune` as discussed. Also added `--dry-run` option to `brew unlinkapps` (many destructive commands have this option, also for symmetry with `brew prune`) and partially rewrote the code from the first two commits.

File: Library/Homebrew/cmd/prune.rb [hunk 3]
```diff
@@ -47,5 +48,7 @@ def prune
       print "and #{d} directories " if d > 0
       puts "from #{HOMEBREW_PREFIX}"
     end unless ARGV.dry_run?
+
+    unlinkapps_prune(:dry_run => ARGV.dry_run?, :quiet => true)
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/prune.rb]: While you're here I wonder if it's worth just adding to `uninstall` too? No worries if you'd rather leave it for another PR.

Comment (UniqMartin) [inline Library/Homebrew/cmd/prune.rb]: I'd prefer to do that in another PR as I expect some discussion. The reason is there are multiple places from where this likely needs to be called, though `uninstall` is the most obvious one.

File: Library/Homebrew/cmd/prune.rb [hunk 4]
```diff
@@ -1,5 +1,6 @@
 require "keg"
 require "cmd/tap"
+require "cmd/unlinkapps"
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/prune.rb]: I don't feel super-strongly about this but we're trying to avoid `cmd/` includes now so if you're feeling extra lovely (can be a future PR) pulling these methods into another module would be :heart:

**Comment by mikemcquaid** [comment]
A few nits but nothing preventing this being merged if you'd rather do the rest in another PR :+1:

File: Library/Homebrew/cmd/prune.rb (hunk 4 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/prune.rb]: You are right and I think it makes more sense to address this here in this PR (big chunks have been rewritten anyway) instead of deferring the code move to a new PR. I'll reshuffle things soon. Thanks for the soft nudge. :heart_eyes_cat:

File: Library/Homebrew/cmd/prune.rb (hunk 3 above)
Comment (mikemcquaid) [inline Library/Homebrew/cmd/prune.rb]: :+1:

File: Library/Homebrew/cmd/prune.rb (hunk 4 above)
Comment (mikemcquaid) [inline Library/Homebrew/cmd/prune.rb]: :+1:

**Comment by mikemcquaid** [comment]
@UniqMartin thoughts on merging as-is?

**Comment by UniqMartin** [comment]
I was hoping to revamp things directly in here to avoid cross `cmd/` requirements, but if you're still fine with it, I'll rebase to resolve the merge conflict and then merge as-is, leaving the other cleanup for later.

**Comment by mikemcquaid** [comment]
@UniqMartin Sounds good 👍

**Comment by UniqMartin** [comment]
Thanks (also for the nudge)! Test failure should be fixed, too. I'll ship as soon as the bot is happy.

**Comment by UniqMartin** [comment]
Closing in favor of Homebrew/brew#15 as this no longer cleanly applies in the new repository.
